File: preinstall/sbr-config/internal/persistence/ifupdown.go
package persistence

import (
	"fmt"
	"log"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/weka/tools/preinstall/sbr-config/internal/constants"
	"github.com/weka/tools/preinstall/sbr-config/internal/exec"
	"github.com/weka/tools/preinstall/sbr-config/internal/models"
)

// IfupdownBackend writes ifupdown configuration for SBR persistence.
// Adds post-up/pre-down lines to interface stanzas in
// /etc/network/interfaces or creates drop-in files in
// /etc/network/interfaces.d/.
type IfupdownBackend struct{}

func (b *IfupdownBackend) WriteConfig(
	interfaces []models.InterfaceInfo,
	tables []models.RoutingTable,
) ([]string, error) {
	tableNum := make(map[string]int)
	for _, t := range tables {
		tableNum[t.Name] = t.Number
	}

	var written []string
	for _, iface := range interfaces {
		tableName := constants.TableNamePrefix + iface.Name
		if _, ok := tableNum[tableName]; !ok {
			continue
		}

		// Build command sets from desired state (not delta).
		upCmds := []string{
			fmt.Sprintf("ip route replace %s dev %s src %s table %s",
				iface.Subnet, iface.Name, iface.IPAddress, tableName),
		}
		if iface.Gateway != nil {
			upCmds = append(upCmds, fmt.Sprintf(
				"ip route replace default via %s dev %s table %s",
				*iface.Gateway, iface.Name, tableName))
		}
		upCmds = append(upCmds, fmt.Sprintf(
			"ip rule add from %s table %s 2>/dev/null",
			iface.IPAddress, tableName))

		downCmds := []string{
			fmt.Sprintf("ip rule del from %s table %s", iface.IPAddress, tableName),
			fmt.Sprintf("ip route flush table %s", tableName),
		}

		// Try to add to existing stanza in interfaces file.
		if b.addToInterfacesFile(iface.Name, upCmds, downCmds) {
			written = append(written, constants.InterfacesFile)
			continue
		}

		// Fall back to drop-in file.
		fpath, err := b.writeDropin(iface, upCmds, downCmds)
		if err != nil {
			log.Printf("[WARN] Failed to write ifupdown drop-in for %s: %v", iface.Name, err)
			continue
		}
		if fpath != "" {
			written = append(written, fpath)
		}
	}

	return written, nil
}

func (b *IfupdownBackend) RemoveConfig() []string {
	var removed []string

	// Remove from interfaces file.
	content, err := exec.ReadFile(constants.InterfacesFile)
	if err == nil && content != "" && strings.Contains(content, constants.ManagedComment) {
		cleaned := removeManagedLines(content)
		if cleaned != content {
			exec.WriteFileAtomic(constants.InterfacesFile, cleaned, 0644)
			removed = append(removed, constants.InterfacesFile)
		}
	}

	// Remove drop-in files.
	entries, err := os.ReadDir(constants.InterfacesDDir)
	if err == nil {
		for _, e := range entries {
			if !strings.HasPrefix(e.Name(), "sbr-") {
				continue
			}
			fpath := filepath.Join(constants.InterfacesDDir, e.Name())
			fc, err := exec.ReadFile(fpath)
			if err != nil || fc == "" {
				continue
			}
			if strings.Contains(fc, constants.ManagedComment) {
				os.Remove(fpath)
				removed = append(removed, fpath)
			}
		}
	}

	return removed
}

func (b *IfupdownBackend) Describe() string {
	return fmt.Sprintf(
		"ifupdown configuration:\n"+
			"  post-up/pre-down lines in %s\n"+
			"  or drop-in files in %s/",
		constants.InterfacesFile, constants.InterfacesDDir)
}

// addToInterfacesFile tries to add post-up/pre-down lines to an existing
// stanza. If managed lines from a previous run already exist, they are
// removed first to avoid accumulating duplicate entries.
// Returns true if successful, false if the stanza wasn't found.
func (b *IfupdownBackend) addToInterfacesFile(
	ifaceName string,
	upCmds, downCmds []string,
) bool {
	content, err := exec.ReadFile(constants.InterfacesFile)
	if err != nil || content == "" {
		return false
	}

	// Remove any existing managed lines first to prevent duplicates.
	if strings.Contains(content, constants.ManagedComment) {
		content = removeManagedLines(content)
	}

	// Find the stanza for this interface.
	pattern := regexp.MustCompile(
		`(?m)^(iface\s+` + regexp.QuoteMeta(ifaceName) + `\s+.*?)` +
			`(?=\niface\s|\nauto\s|\nallow-|\nsource\s|\nmapping\s|\n\z|\z)`)
	loc := pattern.FindStringIndex(content)
	if loc == nil {
		return false
	}
	stanzaEnd := loc[1]

	// Build lines to insert.
	var insertLines []string
	insertLines = append(insertLines, fmt.Sprintf("    %s", constants.ManagedComment))
	for _, cmd := range upCmds {
		insertLines = append(insertLines, fmt.Sprintf("    post-up %s", cmd))
	}
	// Down commands in reverse order.
	for i := len(downCmds) - 1; i >= 0; i-- {
		insertLines = append(insertLines,
			fmt.Sprintf("    pre-down %s 2>/dev/null || true", downCmds[i]))
	}
	insertBlock := strings.Join(insertLines, "\n")

	// Insert at end of stanza.
	newContent := content[:stanzaEnd] + "\n" + insertBlock + content[stanzaEnd:]
	exec.WriteFileAtomic(constants.InterfacesFile, newContent, 0644)
	log.Printf("[INFO] Added SBR lines to %s stanza in %s", ifaceName, constants.InterfacesFile)
	return true
}

// writeDropin writes a drop-in file in /etc/network/interfaces.d/.
func (b *IfupdownBackend) writeDropin(
	iface models.InterfaceInfo,
	upCmds, downCmds []string,
) (string, error) {
	if err := os.MkdirAll(constants.InterfacesDDir, 0755); err != nil {
		return "", fmt.Errorf("create interfaces.d dir: %w", err)
	}

	// Check that interfaces file sources the directory.
	mainContent, _ := exec.ReadFile(constants.InterfacesFile)
	if !strings.Contains(mainContent, "source") && !strings.Contains(mainContent, "interfaces.d") {
		log.Printf("[WARN] %s does not source %s -- drop-in may not be loaded",
			constants.InterfacesFile, constants.InterfacesDDir)
	}

	fpath := filepath.Join(constants.InterfacesDDir, fmt.Sprintf("sbr-%s", iface.Name))

	// Do NOT redefine the interface -- that would conflict with the main
	// stanza. Just add hook commands.
	lines := []string{
		constants.ManagedComment,
		fmt.Sprintf("# Source-based routing hooks for %s (%s)", iface.Name, iface.IPAddress),
		"",
	}
	for _, cmd := range upCmds {
		lines = append(lines, fmt.Sprintf("post-up %s", cmd))
	}
	// Down commands in reverse order.
	for i := len(downCmds) - 1; i >= 0; i-- {
		lines = append(lines, fmt.Sprintf("pre-down %s 2>/dev/null || true", downCmds[i]))
	}
	lines = append(lines, "")

	if err := exec.WriteFileAtomic(fpath, strings.Join(lines, "\n"), 0644); err != nil {
		return "", err
	}
	log.Printf("[INFO] Wrote ifupdown drop-in: %s", fpath)
	return fpath, nil
}

// removeManagedLines removes lines between MANAGED_COMMENT markers.
func removeManagedLines(content string) string {
	lines := strings.Split(content, "\n")
	var newLines []string
	inManagedBlock := false

	for _, line := range lines {
		if strings.Contains(line, constants.ManagedComment) {
			inManagedBlock = true
			continue
		}
		if inManagedBlock {
			trimmed := strings.TrimSpace(line)
			if strings.HasPrefix(trimmed, "post-up") || strings.HasPrefix(trimmed, "pre-down") {
				continue
			}
			inManagedBlock = false
		}
		newLines = append(newLines, line)
	}

	result := strings.Join(newLines, "\n")
	if strings.HasSuffix(content, "\n") && !strings.HasSuffix(result, "\n") {
		result += "\n"
	}
	return result
}

File: preinstall/sbr-config/sbr_config/persistence_backends/ifupdown.py
"""ifupdown (Debian/Ubuntu /etc/network/interfaces) persistence backend."""

import logging
import os
import re
from typing import List

from ..constants import (
    INTERFACES_D_DIR,
    INTERFACES_FILE,
    MANAGED_COMMENT,
    RULE_PRIORITY_INCREMENT,
    RULE_PRIORITY_START,
    TABLE_NAME_PREFIX,
    TABLE_NUMBER_START,
)
from ..models import InterfaceInfo, PlannedChange, RoutingTable
from ..utils import read_file, strip_managed_lines, write_file_atomic
from .base import PersistenceBackend, group_by_name

logger = logging.getLogger(__name__)


class IfupdownBackend(PersistenceBackend):
    """Write ifupdown configuration for SBR persistence.

    Adds post-up/pre-down lines to the interface's existing stanza --
    searching /etc/network/interfaces AND the files in
    /etc/network/interfaces.d/ (stanzas commonly live there via the
    default `source` line). Only when no stanza exists anywhere does it
    fall back to creating its own drop-in; a second stanza for an
    interface that already has one elsewhere would conflict with the
    original's method.
    """

    def write_config(
        self,
        interfaces: List[InterfaceInfo],
        tables: List[RoutingTable],
        changes: List[PlannedChange],
    ) -> List[str]:
        table_num = {t.name: t.number for t in tables}
        written = []

        # Strip managed lines from a previous run ONCE up front, so repeat
        # runs replace them instead of stacking duplicate (and possibly
        # stale) post-up/pre-down blocks. Done before the per-interface
        # loop -- stripping inside it would delete lines just written for
        # an earlier interface in this same run.
        for path in self._stanza_files():
            content = read_file(path)
            if content and MANAGED_COMMENT in content:
                write_file_atomic(path, strip_managed_lines(content))

        for name, entries in group_by_name(interfaces):
            table_name = f"{TABLE_NAME_PREFIX}{name}"
            tnum = table_num.get(table_name)
            if tnum is None:
                continue

            # Derive the full command set from the desired state (not just
            # this run's delta) so persistence is always complete, covering
            # every address on the interface.
            up_cmds = []
            seen_subnets = set()
            for entry in entries:
                if entry.subnet in seen_subnets:
                    continue
                seen_subnets.add(entry.subnet)
                up_cmds.append(
                    f"ip route replace {entry.subnet} dev {name} "
                    f"src {entry.ip_address} table {table_name}"
                )
            gateway = next((e.gateway for e in entries if e.gateway is not None), None)
            if gateway is not None:
                up_cmds.append(
                    f"ip route replace default via {gateway} "
                    f"dev {name} table {table_name}"
                )
            # Explicit priority (planner's allocation scheme, clamped):
            # without it the kernel assigns its own, and the re-added rule
            # would land at a different precedence than the original.
            priority = max(
                RULE_PRIORITY_START,
                RULE_PRIORITY_START
                + (tnum - TABLE_NUMBER_START) * RULE_PRIORITY_INCREMENT,
            )
            for entry in entries:
                up_cmds.append(
                    f"ip rule add from {entry.ip_address} table {table_name} "
                    f"priority {priority} 2>/dev/null"
                )

            down_cmds = [
                f"ip rule del from {entry.ip_address} table {table_name}"
                for entry in entries
            ]
            down_cmds.append(f"ip route flush table {table_name}")

            # Add to the interface's existing stanza wherever it lives --
            # the main interfaces file or a sourced interfaces.d file
            stanza_path = None
            for path in self._stanza_files():
                if self._add_to_stanza_file(path, name, up_cmds, down_cmds):
                    stanza_path = path
                    break
            if stanza_path:
                written.append(stanza_path)
                continue

            # No stanza anywhere: fall back to our own drop-in file
            fpath = self._write_dropin(entries[0], up_cmds, down_cmds)
            if fpath:
                written.append(fpath)

        return written

    def _stanza_files(self) -> List[str]:
        """Files that may hold interface stanzas: the main interfaces
        file first, then sourced interfaces.d files (excluding our own
        sbr-* drop-ins, which are whole-file managed)."""
        paths = []
        if os.path.exists(INTERFACES_FILE):
            paths.append(INTERFACES_FILE)
        if os.path.isdir(INTERFACES_D_DIR):
            for fname in sorted(os.listdir(INTERFACES_D_DIR)):
                if fname.startswith("sbr-"):
                    continue
                fpath = os.path.join(INTERFACES_D_DIR, fname)
                if os.path.isfile(fpath):
                    paths.append(fpath)
        return paths

    def remove_config(self) -> List[str]:
        removed = []

        # Strip managed lines from stanza files (main + user drop-ins)
        for path in self._stanza_files():
            content = read_file(path)
            if content and MANAGED_COMMENT in content:
                cleaned = strip_managed_lines(content)
                if cleaned != content:
                    write_file_atomic(path, cleaned)
                    removed.append(path)

        # Remove our own drop-in files
        if os.path.isdir(INTERFACES_D_DIR):
            for fname in os.listdir(INTERFACES_D_DIR):
                if fname.startswith("sbr-"):
                    fpath = os.path.join(INTERFACES_D_DIR, fname)
                    content = read_file(fpath)
                    if content and MANAGED_COMMENT in content:
                        os.unlink(fpath)
                        removed.append(fpath)

        return removed

    def describe(self) -> str:
        return (
            f"ifupdown configuration:\n"
            f"  post-up/pre-down lines in {INTERFACES_FILE}\n"
            f"  or drop-in files in {INTERFACES_D_DIR}/"
        )

    def _add_to_stanza_file(
        self,
        path: str,
        iface_name: str,
        up_cmds: List[str],
        down_cmds: List[str],
    ) -> bool:
        """Try to add post-up/pre-down lines to an existing stanza in path.

        Returns True if successful, False if the stanza wasn't found.
        """
        content = read_file(path)
        if not content:
            return False

        # Find the stanza for this interface
        pattern = rf'^(iface\s+{re.escape(iface_name)}\s+.*?)(?=\niface\s|\nauto\s|\nallow-|\n\Z|\Z)'
        match = re.search(pattern, content, re.MULTILINE | re.DOTALL)
        if not match:
            return False

        stanza_end = match.end()

        # Build lines to insert
        insert_lines = [f"    {MANAGED_COMMENT}"]
        for cmd in up_cmds:
            insert_lines.append(f"    post-up {cmd}")
        for cmd in reversed(down_cmds):
            insert_lines.append(f"    pre-down {cmd} 2>/dev/null || true")
        insert_block = "\n".join(insert_lines)

        # Insert at end of stanza
        new_content = content[:stanza_end] + "\n" + insert_block + content[stanza_end:]
        write_file_atomic(path, new_content)
        logger.info("Added SBR lines to %s stanza in %s", iface_name, path)
        return True

    def _write_dropin(
        self,
        iface: InterfaceInfo,
        up_cmds: List[str],
        down_cmds: List[str],
    ) -> str:
        """Write a drop-in file in /etc/network/interfaces.d/."""
        if not os.path.isdir(INTERFACES_D_DIR):
            os.makedirs(INTERFACES_D_DIR, exist_ok=True)

        # Check that interfaces file sources the directory
        main_content = read_file(INTERFACES_FILE) or ""
        if "source" not in main_content and "interfaces.d" not in main_content:
            logger.warning(
                "%s does not source %s -- drop-in may not be loaded",
                INTERFACES_FILE,
                INTERFACES_D_DIR,
            )

        fpath = os.path.join(INTERFACES_D_DIR, f"sbr-{iface.name}")

        lines = [
            MANAGED_COMMENT,
            f"# Source-based routing for {iface.name} ({iface.ip_address})",
            "",
            f"auto {iface.name}",
            f"iface {iface.name} inet manual",
        ]
        for cmd in up_cmds:
            lines.append(f"    post-up {cmd}")
        for cmd in reversed(down_cmds):
            lines.append(f"    pre-down {cmd} 2>/dev/null || true")
        lines.append("")

        write_file_atomic(fpath, "\n".join(lines))
        logger.info("Wrote ifupdown drop-in: %s", fpath)
        return fpath
